rules/aip0127/http_template_pattern.go

```diff
@@ -0,0 +1,132 @@
+	pathTemplateToRegex        = strings.NewReplacer("**", resourcePatternAnySegments, "*", resourcePatternSegment)
+)
+
+type resourceReference struct {
+	// The path of the field with the `google.api.resource_reference`. This is
+	// provided as a variable in the HTTPRule.
+	fieldPath string
+	// A template that the resource's pattern string must adhere to. This is
+	// provided by the variable's template in the HTTPRule.
+	pathTemplate string
+	// The name of the resource type. This is used to look up the resource
+	// message.
+	resourceRefName string
+}
+
+// Returns a list of resourceReferences for each variable in all the method's
+// HTTPRule's.
+func resourceRefsForMethod(m *desc.MethodDescriptor) []resourceReference {
+	resourceRefs := []resourceReference{}
+	for _, httpRule := range utils.GetHTTPRules(m) {
+		resourceRefs = append(resourceRefs, resourceRefsForHttpRule(httpRule, m.GetInputType())...)
+	}
+	return resourceRefs
+}
+
+// Returns a resourceReference for every variable in the given HTTPRule.
+func resourceRefsForHttpRule(httpRule *utils.HTTPRule, msg *desc.MessageDescriptor) []resourceReference {
+	resourceRefs := []resourceReference{}
+	for fieldPath, template := range httpRule.GetVariables() {
+		// Find the (sub-)field in the message corresponding to the variable's
+		// field path.
+		if resourceRefField := utils.FindFieldDotNotation(msg, fieldPath); resourceRefField != nil {
+			// Extract the name of the resource referenced by this field.
+			if resourceRef := utils.GetResourceReference(resourceRefField); resourceRef != nil {
+				// TODO(acamadeo): Support the case where a resource has
+				// multiple parent resources.
+				if resourceRef.GetChildType() != "" {
+					continue
+				}
+
+				res := resourceReference{fieldPath: fieldPath, pathTemplate: template, resourceRefName: resourceRef.GetType()}
+				resourceRefs = append(resourceRefs, res)
+			}
+		}
+	}
+	return resourceRefs
+}
+
+// Constructs a regex from the HTTPRule's path template representing resource
+// patterns that it will match against.
+func compilePathTemplateRegex(pathTemplate string) (*regexp.Regexp, error) {
+	pattern := fmt.Sprintf("^%s$", pathTemplateToRegex.Replace(pathTemplate))
+	return regexp.Compile(pattern)
+}
+
+func anyStringsMatchRegex(regex *regexp.Regexp, strs []string) bool {
+	for _, str := range strs {
+		if regex.MatchString(str) {
+			return true
+		}
+	}
+	return false
+}
+
+// Checks whether the HTTP pattern specified in `resourceRef` matches any of the
+// patterns defined for that resource.
+func checkHttpPatternMatchesResource(m *desc.MethodDescriptor, resourceRef resourceReference) []lint.Problem {
+	annotation := utils.FindResource(resourceRef.resourceRefName, m.GetFile())
+	if annotation == nil {
+		message := fmt.Sprintf("Unable to find resource with name %q", resourceRef.resourceRefName)
+		return []lint.Problem{{Message: message, Descriptor: m.GetInputType()}}
+	}
```

Comment:
Ok, so, I know this might sound weird BUT...we should just return `nil` here.

I think that if a `google.api.resource_reference` is unresolvable, it should be warned on by a dedicated rule. This AIP and this rule do not deal in resolution of resource references, so we need to keep that functionality separate.

Reply:
Makes total sense, thanks!